### PR TITLE
Rework Table for use with multiple types

### LINKED ISSUE #1648: Log every action that occurs as the result of a command issued to stratisd at the info level
Currently, users interact w/ the stratisd D-Bus layer. Every engine action that could occur as a result of a command issued on the D-Bus should be logged. However, this logging should not occurs at the D-Bus layer itself, but in a hypothetical layer between the D-Bus layer and the engine.

### LINKED ISSUE #1877: Need to distinguish PoolUuid, DevUuid, FilesystemUuid by something more than aliases
Currently, all these are just aliases for Uuid. This will make some definitions necessary for https://github.com/stratis-storage/stratisd/pull/1874 to have conflicts like:
```
error[E0119]: conflicting implementations of trait `std::fmt::Display` for type `engine::types::actions::RenameAction<uuid::Uuid>`:
  --> src/engine/types/actions.rs:33:1
   |
27 | impl fmt::Display for RenameAction<PoolUuid> {
   | -------------------------------------------- first implementation here
...
33 | impl fmt::Display for RenameAction<DevUuid> {
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ conflicting implementation for `engine::types::actions::RenameAction<uuid::Uuid>`

error: aborting due to previous error
```

### PR DESCRIPTION
**Pull request by jbaublitz**
Supersedes #2321 
Closes #1648
Closes #1877

### CODE REVIEW

**Comment by mulkieran** [comment]
@jbaublitz Could you rebase and resolve the minor conflicts? We will aim to get this in in the next release, if at all possible.

**Comment by jbaublitz** [comment]
Rebased!

**Review by mulkieran — CHANGES_REQUESTED** [review]
It looks good to me so far. But looking back I remember it was all in aid of making this work: https://github.com/stratis-storage/stratisd/pull/2288/files. Right here: https://github.com/stratis-storage/stratisd/pull/2288/files was where we realized that we wanted ```PoolUuid``` to be its own type, not just a type alias, so that we could do similar but different impls of ```Display``` for ```CreateAction<FilesystemUuid>``` and ```CreateAction<DevUuid>``` for example. And the problem was that if we did that, we'ld have to implement some methods on each of these types, like ```to_simple_ref()``` etc.

I'm afraid you'll have to remind me of what the plan was to solve that problem.

File: src/engine/strat_engine/cleanup.rs [hunk 1]
```diff
@@ -0,0 +1,28 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
```
[the remark this answers is not in the document]
Comment (jbaublitz) [reply src/engine/strat_engine/cleanup.rs:1]: Whoops! I'll remove this.

**Comment by jbaublitz** [comment]
I think the plan to solve this was to implement `AsUuid` for all of the UUID wrapped types and simply call `.as_uuid()` on those structs and use the existing methods. Is that still what you want to do?

**Comment by mulkieran** [comment]
> I think the plan to solve this was to implement `AsUuid` for all of the UUID wrapped types and simply call `.as_uuid()` on those structs and use the existing methods. Is that still what you want to do?

That would be fine w/ me. The other way is to use default methods in the AsUuid trait for the few methods that we actually use, which might be less intrusive. So, something like:

```
trait AsUuid {
  fn as_uuid(&self) -> &Uuid;

  fn to_simple_ref(&self) -> SimpleRef {
    self.as_uuid().to_simple_ref()
  }

  // any other funcs, hopefully few
}

#[derive(...)]
struct PoolUuid(Uuid);

impl AsUuid for PoolUuid {
  fn as_uuid(&self) -> &Uuud {
    &self.0
 }
}
```

**Comment by mulkieran** [comment]
We also might use Deref, or possibly a trait bounded by Deref.

File: src/engine/strat_engine/cleanup.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (jbaublitz) [reply src/engine/strat_engine/cleanup.rs:1]: Resolved!

**Comment by jbaublitz** [comment]
@mulkieran I found the bug. It boils down to the XFS CLI expecting dashes in the UUID while our output on disk is without dashes. This raises an interesting question about the `Display` trait for our wrapper types. Should those display the output with dashes and we can continue calling `.to_simple_ref()` if we need it without dashes? That seems consistent with what we previously did and maybe it's best to leave it that way.

**Comment by mulkieran** [comment]
> @mulkieran I found the bug. It boils down to the XFS CLI expecting dashes in the UUID while our output on disk is without dashes. This raises an interesting question about the `Display` trait for our wrapper types. Should those display the output with dashes and we can continue calling `.to_simple_ref()` if we need it without dashes? That seems consistent with what we previously did and maybe it's best to leave it that way.

I think that's the correct decision.

**Comment by jbaublitz** [comment]
Okay, then I think this is ready for review. Moving.

**Review by mulkieran — CHANGES_REQUESTED** [review]
Looks great! Just two requests...

Do you want to merge this PR as soon as it's ready, or do you want to add the macro work on to this PR as a separate commit?

**Comment by jbaublitz** [comment]
> Looks great! Just two requests...
> 
> Do you want to merge this PR as soon as it's ready, or do you want to add the macro work on to this PR as a separate commit?

I think the macro work as a separate commit might be for the best just in case I need to change anything.

**Comment by mulkieran** [comment]
> > Looks great! Just two requests...
> > Do you want to merge this PR as soon as it's ready, or do you want to add the macro work on to this PR as a separate commit?
> 
> I think the macro work as a separate commit might be for the best just in case I need to change anything.

Sounds good.

File: src/dbus_api/pool/pool_2_0/methods.rs [hunk 2]
```diff
@@ -113,17 +119,26 @@ pub fn destroy_filesystems(m: &MethodInfo<MTFn<TData>, TData>) -> MethodResult {
...
     let mut engine = dbus_context.engine.borrow_mut();
     let (pool_name, pool) = get_mut_pool!(engine; pool_uuid; default_return; return_message);
 
     let filesystem_map: HashMap<FilesystemUuid, dbus::Path<'static>> = filesystems
         .filter_map(|path| {
             m.tree.get(&path).and_then(|op| {
-                op.get_data()
-                    .as_ref()
-                    .map(|d| (d.uuid, op.get_name().clone()))
+                op.get_data().as_ref().and_then(|d| match d.uuid {
```
[the remark this answers is not in the document]
Comment (jbaublitz) [reply src/dbus_api/pool/pool_2_0/methods.rs]: Resolved.

File: src/dbus_api/api/manager_2_0/methods.rs [hunk 3]
```diff
@@ -36,9 +36,16 @@ pub fn destroy_pool(m: &MethodInfo<MTFn<TData>, TData>) -> MethodResult {
         .tree
         .get(&pool_path)
         .and_then(|op| op.get_data().as_ref())
-        .map(|d| d.uuid)
+        .map(|d| &d.uuid)
     {
-        Some(uuid) => uuid,
+        Some(StratisUuid::Pool(uuid)) => *uuid,
```
[the remark this answers is not in the document]
Comment (jbaublitz) [reply src/dbus_api/api/manager_2_0/methods.rs]: Resolved.

**Comment by mulkieran** [comment]
@jbaublitz Looks good! Please carry on with macro work.

**Comment by jbaublitz** [comment]
@mulkieran The prototype is done. I have some suggestions for how we can add additional logging information based on what I saw of the return types contained inside the data structures like `CreateAction` for some methods. You can already see an example of what I'm suggesting in the `impl` for `impl Display for DeleteAction<Key>` vs. `impl Display for DeleteAction<Clevis>`.

@drckeefe Take a look at the logs and see what you think.

**Comment by mulkieran** [comment]
We're still not getting the output we wanted...that must be because the env variable isn't being properly observed. I'm going to try to come up with a fix for that in stratis-cli.

**Comment by mulkieran** [comment]
> We're still not getting the output we wanted...that must be because the env variable isn't being properly observed. I'm going to try to come up with a fix for that in stratis-cli.

cli-with-stratisd test is all ready now. @drckeefe @bgurney-rh please take a look.

File: src/engine/engine.rs [hunk 4]
```diff
@@ -53,15 +53,15 @@ pub trait KeyActions {
         &mut self,
         key_desc: &KeyDescription,
         key_fd: RawFd,
-    ) -> StratisResult<MappingCreateAction<()>>;
+    ) -> StratisResult<MappingCreateAction<Key>>;
```
Comment (mulkieran) [inline src/engine/engine.rs:56]: Need to fix up header docs to match.

**Review by bgurney-rh — APPROVED** [review]

**Review by mulkieran — CHANGES_REQUESTED** [review]
Noticed one thing...design looks great to me.

File: src/engine/engine.rs (hunk 4 above)
Comment (jbaublitz) [reply src/engine/engine.rs:56]: Resolved.

**Review by drckeefe — APPROVED** [review]

**Comment by mulkieran** [comment]
Ok! In standup we decided that we will format UUIDs in log messages and in the result of the engine state report with hyphens. So the changes currently in this PR will remain and some to_simple_ref() calls will have to be removed in various places.

**Comment by jbaublitz** [comment]
@mulkieran Okay I think this is all set.

**Review by mulkieran — CHANGES_REQUESTED** [review]
There's one thing I forgot...can you remove ```name``` and ```uuid``` from ```TestThing``` definition. I think they exist because they were a mandatory part of the value object in the original Table struct. But since they aren't anymore, and aren't used to test for anything anywhere, they should be removed.

**Comment by jbaublitz** [comment]
@mulkieran Okay this should be ready.

**Review by mulkieran — CHANGES_REQUESTED** [review]
Looks good. Please close this and open a new PR for final review.

**Pull request closed without merging**